analysis/generate_dummy_rows.py

```diff
@@ -0,0 +1,59 @@
+"""Generate dummy rows to match the query in analysis/query.sql.
+
+opensafely exec python:latest python -m analysis.generate_dummy_rows
+
+"""
+import functools
+
+import pandas
+from numpy import random
+
+from analysis import utils
+
+
+rng = random.default_rng(seed=1)
+
+
+def main():
+    table_names = [
+        "APCS",
+        "Appointment",
+        "CodedEvent",
+        "CPNS",
+        "EC",
+        "ICNARC",
+        "ONS_Deaths",
+        "OPA",
+        "SGSS_AllTests_Negative",
+        "SGSS_AllTests_Positive",
+        "SGSS_Negative",
+        "SGSS_Positive",
+    ]
+    from_date = "2016-01-01"
+    to_date = "2022-12-31"
+    f_out = utils.OUTPUT_DIR / "generate_dummy_rows" / "dummy_rows.csv.gz"
+
+    utils.makedirs(f_out.parent)
+    data_frame = make_dummy_rows(table_names, from_date, to_date)
+    data_frame.to_csv(f_out, index=False)
+
+
+@functools.singledispatch
```

Comment:
Is single dispatch necessary here?

Reply:
We have two functions that return a data frame of dummy rows; they differ in whether they return this for one table or for several tables. We could name each function differently, e.g. `make_dummy_rows_for_one_table` and `make_dummy_rows_for_several_tables`, but that is inelegant. Alternatively, we could use `functools.singledispatch` to determine which implementation is called: pass a `str` and get a data frame of dummy rows for one table; pass a `list` and get a data frame of dummy rows for several tables, one for each item in the list. That is more elegant; or less inelegant 🙂 

Of course, we could merge the two functions into one. However, then we'd have to store a list of data frames (one entry for each item in `table_names`) that we'd then pass to `pandas.concat`. My sense is that `pandas.concat` copies the input data frames, so we'd end up holding the list of data frames and the concatenated data frame in memory at once. Does that matter? Probably not. The concatenated data frame is less than a megabyte. I don't think we want to encourage this pattern, though.

Maybe we don't have to store a list of data frames 🤔 Maybe we could use an inner (generator) function? We could then remove `functools.singledispatch` and avoid copying the input data frames.

So, I don't think `functools.singledispatch` is necessary. But I'm not sure whether the alternatives are better, worse, or just different. What are your thoughts?

Reply:
My main argument against it would be that using `@singledispatch` makes it harder for people new to Python to understand the code, and without any real benefit over using two functions with different names.  (Elegance is in the eye of the beholder, but to me having multiple functions with similar names is fine and not inelegant.)

One other thought: it's really not worth worrying about concatenating a few megabytes here and there, especially if it changes how you write code.

Reply:
I'm less keen on multiple functions with similar names: each means an extra unit of cognition, whereas overloading doesn't. (Yes, my justification has changed from elegance to cognitive effort 🙂) However, "people new to Python" is a group I'd like to address with this code; even if they don't modify it, then they may end up using the pattern. And if they _do_ end up using the pattern, then they may end up using it with much larger data frames (e.g. concatenating multiple datasets, each for different sub-populations). For this reason, I'm still keen to avoid storing a list of data frames.

So, `make_dummy_rows` now encloses an inner function, which returns a data frame of dummy rows for one table.